jumanji/env.py

```diff
@@ -33,9 +33,10 @@ class StateProtocol(Protocol):
 
 
 State = TypeVar("State", bound="StateProtocol")
+ActionSpec = TypeVar("ActionSpec", bound=specs.Array)
```

Comment:
I think this adds unnecessary complexity to the code without serving a downstream purpose. Unless I'm mistaken, we don't seem to use `ActionSpec` for anything practical. Therefore, I would suggest removing the type variable and keeping the typing `spec.DiscreteArray` or whatever the environments have. What do you think?

Reply:
The purpose of `ActionSpec` is to type the `Environment.action_spec` method. Each environment class shouldn't need to override this method, so it needs to be typed correctly on the base `Environment` class. This requires a generic. The only other solution I can see would be to implement the following on each environment.
```
@property
def action_spec(self) -> specs.Whatever:
    return self._action_spec
```

Reply:
Hmm I see. In this PR, it seems we are removing typing for the `observation_spec` method since we are moving it to the `Environment` base class. And we are doing similarly for the `action_spec`method but considering a `TypeVar` to keep some typing. Since the observation and action specs play similar roles as being environment-specific, I would argue for a similar implementation between the two.
- Either we define both `observation_spec` and `action_spec` in each environment class (allowing for environment-specific `spec.Spec[Observation]` and e.g. `spec.DiscreteArray` definitions).
- Or we just define the `make_observation_spec` and `make_action_spec` in each environment with a TypeVar-like typing.

I think solution 1 is clearer imo

Reply:
I agree that `observation_spec` and `action_spec` should probably have a similar implementation. However, I think solution 2 would be better. While it does lead to some unwieldy class definitions, it would also be a step towards addressing issue #221 with wrapper typing.

Reply:
I think I agree with @aar65537 on this. Personally prefer the TypeVar approach, as it allows us to not redefine `observation_spec` and `action_spec` in each environment. But we should definitely add a type for for the obs spec.

I have a couple concerns though. We may start having too many type vars in environment definitions e.g

```class Game2048(Environment[State, specs.DiscreteArray, specs.BoundedArray]):```

Which can be confusing to environment authors as to which generic refers to which spec, especially if we find the need to add more generics in the future.

Also are nested generics possible, could we ever do:

```class Game2048(Environment[State, specs.DiscreteArray, specs.Spec[Observation]]):```

I don't think they are, especially not in python 3.9/3.10. If this isn't possible I think we lose a lot more typing, so then I'd prefer the bloat of having both `observation_spec` and `make_observation_spec` methods in all classes for proper type definitions. 

A possible way around this is to have an `Observation` typevar and then type `observation_spec` in the base class as follows:

```
@property
def observation_spec() -> Specs.Spec[Observation]
```

Can we assume that this is always the type of observation spec?

(Thanks for this great contribution @aar65537! :fire:)

Reply:
I just updated the branch with the typevar approach and also fixed the issue with the wrappers. I share your concern with the number of typevars in the environment definitions, but I really don't see a better way.

Nested generics are valid, but they don't provide meaningful nesting. You can't access the inner type (`Observation` in this case) from the class. For this reason, I chose to use an `Observation` typevar and just assume `observation_spec` returns a `specs.Spec[Observation]`. This is true for all current environments, and in my opinion is a reasonable assumption. 

The only issue is some type information can be lost if your observation spec is a more specific type. For example, `FakeEnvironment` returns a `specs.Array` as its observation spec, but it is typed as `specs.Spec[chex.Array]`. `specs.Array` is a subtype of `specs.Spec[chex.Array]`, so there is no error, but the typing is not as precise as it could be.

Reply:
I think these assumptions are fine, I'm happy to merge this as is, @clement-bonnet what are your thoughts?